diesel/src/pg/expression/functions.rs

```diff
@@ -2078,3 +2078,100 @@ define_sql_function! {
 
     fn jsonb_array_length<E: JsonbOrNullableJsonb + MaybeNullableValue<Integer>>(jsonb: E) -> E::Out;
 }
+
+#[cfg(feature = "postgres_backend")]
+define_sql_function! {
+    /// Returns target with new_value inserted, if the item designated by the path is an array element.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # include!("../../doctest_setup.rs");
+    /// #
+    /// # fn main() {
+    /// #     run_test().unwrap();
+    /// # }
+    /// #
+    /// # fn run_test() -> QueryResult<()> {
+    /// #     use diesel::dsl::jsonb_insert;
+    /// #     use diesel::sql_types::{Array, Jsonb, Nullable, Text};
+    /// #     use serde_json::{Value, json};
+    /// #     let connection = &mut establish_connection();
+    /// let result = diesel::select(jsonb_insert::<Jsonb, Array<Text>, _, _, _>(json!([1, 2, 3]), vec!["2"], json!(1)))
+    ///                 .get_result::<Value>(connection)?;
+    /// assert_eq!(json!([1, 2, 1, 3]), result);
+    ///
+    /// let result = diesel::select(jsonb_insert::<Jsonb, Array<Text>, _, _, _>(json!([1, 2, 3]), Vec::<String>::new(), json!(1)))
+    ///                 .get_result::<Value>(connection)?;
+    /// assert_eq!(json!([1, 2, 3]), result);
+    ///
+    /// let result = diesel::select(jsonb_insert::<Nullable<Jsonb>, Array<Text>, _, _, _>(Option::<Value>::None, vec!["1"], json!(1)))
+    ///                 .get_result::<Option<Value>>(connection)?;
+    /// assert_eq!(None, result);
+    ///
+    /// let result = diesel::select(jsonb_insert::<Jsonb, Nullable<Array<Text>>, _, _, _>(json!([1, 2, 3]), None::<Vec<String>>, json!(1)))
+    ///                 .get_result::<Option<Value>>(connection)?;
+    /// assert_eq!(None, result);
+    ///
+    /// let result = diesel::select(jsonb_insert::<Nullable<Jsonb>, Array<Text>, _, _, _>(json!([1, 2, 3]), vec!["1"], Option::<Value>::None))
+    ///                 .get_result::<Option<Value>>(connection)?;
+    /// assert_eq!(None, result);
+    ///
+    ///
+    /// #     Ok(())
+    /// # }
+    /// ```
+    fn jsonb_insert<J: JsonbOrNullableJsonb + SingleValue, P: TextArrayOrNullableTextArray + CombinedNullableValue<J, Jsonb>>(target: J, path: P, value: J) -> P::Out;
```

Comment:
We need to have separate generic types for the first and last argument here to allow using different types in these places. That allows us to use a column in one place and an expression in another place while calling this function. That means we won't be able to use `CombineNullableValue` as it is, we might need another helper trait there or some more experimentation how to combine three or four values with the existing trait.